ramips: add support for TP-Link Archer C5 v4

namiltd opened the pull request:
Patch https://github.com/openwrt/openwrt/pull/14586 successor. PR https://github.com/openwrt/openwrt/pull/14314 compatible with main branch.

TP-Link Archer C5 v4 is a dual band router with 5 GbE ports
Advertised as AC1200 for its 867Mbps (2x2) 5GHz band
and 300 Mbps (2x2) 2.4GHz band.

Specs:
- SoC: MediaTek MT7620A
- Ethernet: 5x GbE ports (Realtek RTL8367S)
- Wireless 2.4GHz: MediaTek MT7620A
- Wireless 5GHz: MediaTek MT7612E
- RAM: 64MiB
- ROM: 8MiB (GD25Q64CSIG)
- 1 USB 2.0 port
- 2 Buttons (WPS and reset)
- 8 LEDs

Currently, you can install OpenWrt via serial console or TFTP.

Flash instructions via serial console:
1. Rename the factory.bin to to test.bin
2. start a TFTP server from IP address 192.168.0.225 and serve the image named test.bin
3. connect your device to the LAN port
4. power up the router and press 4 on the console to stop the boot process.
5. enter the following commands on the router console
	tftp 0x80060000 test.bin
	erase tplink 0x20000 0x7a0000
	cp.b 0x80060000 0x20000 0x7a0000
	reset

Flash instructions via TFTP:
1. Update orginal firmware of the router to the latest one.
2. Rename openwrt-ramips-mt7620-tplink_archer-c5-v4-squashfs-tftp-recovery.bin to tp_recovery.bin
3. Change computer IP to 192.168.0.66
4. Run TFTP serwer
5. Start the router with the reset button pressed, the file will be automatically downloaded and after a while the router will restart.
6. After updating, set your computer's IP to DHCP

A little note: Since there are problems with the performance of 2.4GHz WiFi for Mediatek MT7620 with enabled 802.11w Management Frame Protection, use WPA2-PSK security. For 5GHz this problem does not occur, so you can enable WPA3-SAE or WPA2-PSK/WPA3-SAE security mode.

Additionally, limit the WiFi power to that allowed in your country because OpenWrt does not take into account the antenna gain for this router (5dBi).

luizluca commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,114 @@
+		};
+
+		wlan2g {
+			label = "green:wlan2g";
```
Led labels were replaced by function and color. See any other dts.

namiltd replied:
But not yet for mt7620a_tplink_archer.dtsi and its successors. In my opinion, the entire group should be improved at the same time, but this should be done in a separate PR.

namiltd replied:
Right, I just did it at the same stage as the other files in the mt7620a_tplink_archer group.

luizluca commented:
The main obstacle for this PR is the switch change. It is a major change (as it might affect many devices) behind a new device PR. I recommend to send that in an isolated PR to bring the required attention. You can keep this PR as it is just mentioning it requires the other PR. If that one gets merged, this PR will automatically shrinks.

The first time I worked in with a PR like this, I'm was including the switch change as it was. Now, I have enough experience to review the switch change, although I don't have the commit access to finally merge it.

abajk commented:
Is there something that blocks the use of DSA driver in this device?

namiltd commented:
DSA support is built into the kernel of the 6.1 kernel tested in the main branch. You can find an implementation attempt here: https://github.com/luizluca/openwrt/commits/ramips-6-1+namiltd+dsa
This driver does not have DSA but is tested and working.

luizluca commented:
> DSA support is built into the kernel of the 6.1 kernel tested in the main branch. You can find an implementation attempt here: https://github.com/luizluca/openwrt/commits/ramips-6-1+namiltd+dsa This driver does not have DSA but is tested and working.

That branch is a moving target. I'm using it as the platform to work on the upstream Realtek DSA driver. It would be good to have those improvements merged upstream and backported before the first OpenWrt device starts to use that DSA driver.

The kernel 6.1 for ramips was just merged. I need to rebase and retest c5-v4 against it. On my device, that branch is crashing when it initializes the PCI device (wireless) or when I try to sysupgrade while not in failsafe mode. I burned my serial port and maybe something else along the way, so I need to make sure it is not related to that.

There are a couple of limitations with the use of the current DSA driver. The first one is the lack of checksum offloading, as the SoC offloading engine cannot handle packets with an alien (Realtek) CPU tag. So, any traffic from the router will require a little extra processing to calculate the checksums in software.

A second limitation is from the current DSA driver as it does not support switch forwarding offload. Thus, any forwarding between two Ethernet ports will always pass through the CPU, requiring again more CPU processing. This problem can be fixed by improving the DSA driver, and I implemented it last year. I used SVL (Shared VLAN Learning) but the upstream maintainer rejected it, saying that he had an almost working IVL implementation (that never came to light). I plan to rebase my SVL implementation in the c5v4 branch once other changes get merged upstream.

There are a couple of changes to the ramips Ethernet and internal switch to make it compatible with the current Realtek DSA driver. First, it needs to support the extra bytes to fit the CPU tag. It also needs to be able to disable the internal switch learning, as all traffic must reach the CPU to be forwarded (the second limitation I mentioned).

In the end, I still see a reason to use swconfig until, at least, the DSA driver gets forwarding offload. The complexity of changes that might affect other devices in both cases are similar (improving swconfig rtl8367b driver or improving ramips/mt7620 eth/switch). This device is getting old, and we need to have it supported sooner rather than later.

The ramips/mt7620 might never see DSA support as a whole. Although the switch is supported upstream, the ethernet driver only supports mt7621. I don't think someone will invest such time to improve a SoC that only supports 100M switching (when not paired with an extra switch).

DragonBluep commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,124 @@
+
+&ethernet {
+	pinctrl-names = "default";
+	pinctrl-0 = <&rgmii1_pins &mdio_pins>;
```
```suggestion
	pinctrl-0 = <&rgmii1_pins>, <&mdio_pins>;
```

namiltd replied:
The vast majority of dts files in this folder have the same syntax as in my file, which is what I suggested.

DragonBluep commented on target/linux/ramips/mt7620/base-files/etc/board.d/02_network:
```diff
@@ -235,6 +235,12 @@ ramips_setup_interfaces()
+		ucidef_add_switch "switch0"
+		ucidef_add_switch_attr "switch0" "enable" "0"
+		ucidef_add_switch "switch1" \
+			"0:lan:4" "1:lan:3" "2:lan:2" "3:lan:1" "4:wan" "7@eth0"
```
Does MT7620 OpenWrt really support port 7? It seems all other devices use port 6.

namiltd replied:
Yes, port 7 is correct.

DragonBluep commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,124 @@
+	pinctrl-0 = <&rgmii1_pins &mdio_pins>;
+
+	nvmem-cells = <&macaddr_rom_f100>;
+	nvmem-cell-names = "mac-address";
```
Already exists in `mt7620a_tplink_archer.dtsi`.

DragonBluep commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,124 @@
+	pinctrl-0 = <&pa_pins>;
+	pinctrl-1 = <&pa_gpio_pins>;
+
+	ralink,mtd-eeprom = <&radio 0x0>;
```
```suggestion
```

namiltd replied:
Done

DragonBluep commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,124 @@
+
+	ralink,mtd-eeprom = <&radio 0x0>;
+	nvmem-cells = <&macaddr_rom_f100>;
+	nvmem-cell-names = "mac-address";
```
```
	nvmem-cells = <&eeprom_radio_0>, <&macaddr_rom_f100 0>;
	nvmem-cell-names = "eeprom", "mac-address";
```

namiltd replied:
Fixed

DragonBluep commented on target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:
```diff
@@ -0,0 +1,124 @@
+	};
+
+	rtl8367s {
+		compatible = "realtek,rtl8367c";
```
The local driver doesn't support `realtek,rtl8367c`.

namiltd replied:
Fixed

namiltd committed:
ramips: fix TP-Link Archer MAC address for ethernet

This fixes:
[    0.981884] OF: /ethernet@10100000: #nvmem-cell-cells = 1 found 0

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd committed:
ramips: add support for TP-Link Archer C5 v4

TP-Link Archer C5 v4 is a dual band router with 5 GbE ports
Advertised as AC1200 for its 867Mbps (2x2) 5GHz band
and 300 Mbps (2x2) 2.4GHz band.

Specs:
- SoC: MediaTek MT7620A
- Ethernet: 5x GbE ports (Realtek RTL8367S)
- Wireless 2.4GHz: MediaTek MT7620A
- Wireless 5GHz: MediaTek MT7612E
- RAM: 64MiB
- ROM: 8MiB (GD25Q64CSIG)
- 1 USB 2.0 port
- 2 Buttons (WPS and reset)
- 8 LEDs

Flash instructions:
Currently one has to install OpenWrt only via the serial console

1. Rename the factory.bin to to test.bin
2. start a TFTP server from IP address 192.168.0.225 and serve the image named test.bin
3. connect your device to the LAN port
4. power up the router and press 4 on the console to stop the boot process.
5. enter the following commands on the router console
	tftp 0x80060000 test.bin
	erase tplink 0x20000 0x7a0000
	cp.b 0x80060000 0x20000 0x7a0000
	reset

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>
[Update leds, add fast-read]
Signed-off-by: Gaspare Bruno <gaspare@anlix.io>
[Rebuilt version based on mt7620 tplink_archer.dtsi, support for external LNA, remove bad cell count info]
Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

luizluca commented:
@namiltd,  [this](https://github.com/luizluca/openwrt/tree/c5v4_dsa) is my current DSA branch based on this PR (rebased on current main to get ramips kernel 6.1).

Both this PR and my branch work as expected with:

```
CONFIG_TARGET_ramips=y
CONFIG_TARGET_ramips_mt7620=y
CONFIG_TARGET_ramips_mt7620_DEVICE_tplink_archer-c5-v4=y
CONFIG_TESTING_KERNEL=y
```

rmilecki commented:
`mt7620a_tplink_archer-c5-v4.dts` misses `SPDX` header.

@namiltd: can you open pull request adding it? Ideally
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
or
```
// SPDX-License-Identifier: GPL-2.0-only OR MIT
```

@hauke: please kindly try to verify DTS licensing before merging

namiltd commented:
@rmilecki: https://github.com/openwrt/openwrt/pull/14718
